### PR TITLE
Build tree and get merkle_root from hashed inputs  (#205)

### LINKED ISSUE #205: Implement `merkle_root`
## Description

Implement the `merkle_root` function from the spec.

This was split-off from the #104 mega-issue.

### PR DESCRIPTION
**Pull request by johnsBeharry**
## Issue Addressed

Resolves #205 

## Proposed Changes

- Implement the merkle_root reference [from the spec](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#merkle_root) in Rust

## Additional Info

1. I was having some issues splitting the function into it's own file so I've opted to have it remain in the `lib.rs` together with `hash()`.

### CODE REVIEW

**Comment by paulhauner** [comment]
Woo! :smile: Looking good, from a cursory glance.

The tests are failing on [`rustfmt`](https://github.com/rust-lang/rustfmt) -- we check that all code formatting passes `rustfmt`. Quickest way to fix this is to run `cargo fmt --all`.

It's worth setting up your IDE to always run `rustfmt` when you save a file. It saves these errors and it saves a bunch of time trying to format stuff.

Ping me on Gitter when the tests pass and I'll do a review.

**Comment by johnsBeharry** [comment]
@paulhauner seems like my comments were too short, there is another fmt error in protos though - 

This branch should be up to date with master:head :thinking: 

```
→ cargo fmt --all -- --check
error[E0583]: file not found for module `services`
 --> /home/johns/ethereum/lighthouse/protos/src/lib.rs:3:9
  |
3 | pub mod services;
  |         ^^^^^^^^
  |
  = help: name the file either services.rs or services/mod.rs inside the directory "/home/johns/ethereum/lighthouse/protos/src"

Error writing files: parse error
```

**Comment by johnsBeharry** [comment]
@paulhauner tests seems to have passed now quite odd.

File: eth2/utils/hashing/src/lib.rs [hunk 1]
```diff
@@ -8,6 +8,31 @@ pub fn hash(input: &[u8]) -> Vec<u8> {
     result
 }
 
+// Get merkle root of some hashed values - the input leaf nodes is expected to already be hashed
```
Comment (paulhauner) [inline eth2/utils/hashing/src/lib.rs]: If we make these triple-slashes (`///`) they'll show up in Rust documentation.

File: eth2/utils/hashing/src/lib.rs [hunk 2]
```diff
@@ -8,6 +8,31 @@ pub fn hash(input: &[u8]) -> Vec<u8> {
...
+// Outputs a `Vec<u8>` byte array of the merkle root given a set of leaf node values.
+pub fn merkle_root(values: &[Vec<u8>]) -> Vec<u8> {
```
Comment (paulhauner) [inline eth2/utils/hashing/src/lib.rs]: There are a few occasions where this algorithm won't function:

- When `values.len() == 0` the function will panic on line 33.
- When `values.len().is_power_of_two() == false` the function will produce an invalid result. (See the comment [here](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#merkle_root))

There's two ways we can deal with this:
1. Add some explicit `assert` statements to make it clear that these are invalid inputs (e.g., `assert!(values.len().is_power_of_two())`
2. Change the return type of the function to `Option<Vec<u8>>` and return `None` if `values.is_empty()` or `!values.len().is_power_of_two()` (and change line 33 to `Some(o[1].clone())`).

I think (2) is the safest in this scenario, so I'd recommend to go ahead with that.

Apologies, I didn't consider this last time we were chatting.

**Comment by paulhauner** [comment]
Hey @johnsBeharry -- could you please sign the CLA and merge in master :)

**Comment by paulhauner** [comment]
I'm going to ignore Jenkins here (it's getting phased out) and merge in.

Thanks @johnsBeharry!

**Comment by spble** [comment]
> I'm going to ignore Jenkins here (it's getting phased out) and merge in.

Yep, the error on Jenkins was:
`Error response from daemon: Error processing tar file(exit status 1) ... no space left on device`
So no issues with code anyway.

**Pull request merged**